pytest/node_plugin.py

```diff
@@ -0,0 +1,293 @@
+"""Pytest plugin implementing a Node fixture for running remote commands."""
+from __future__ import annotations
+
+import json
+import logging
+import platform
+import typing
+from io import BytesIO
+from uuid import uuid4
+
+import fabric  # type: ignore
+import invoke  # type: ignore
+from fabric import Connection
+from invoke import Context
+from invoke.runners import Result  # type: ignore
+from tenacity import retry, stop_after_attempt, wait_exponential  # type: ignore
+
+import pytest
+
+if typing.TYPE_CHECKING:
+    from typing import Any, Dict, Iterator, Optional, Tuple
+
+    from _pytest.fixtures import FixtureRequest
+
+# Setup a sane configuration for local and remote commands. Note that
+# the defaults between Fabric and Invoke are different, so we use
+# their Config classes explicitly.
+config = {
+    "run": {
+        # Show each command as its run.
+        "echo": True,
+        # Disable stdin forwarding.
+        "in_stream": False,
+        # Don’t let remote commands take longer than five minutes
+        # (unless later overridden). This is to prevent hangs.
+        "command_timeout": 1200,
+    }
+}
+
+
+# Provide a configured local Invoke context for running commands
+# before establishing a connection. (Use like `local.run(...)`).
+invoke_config = invoke.Config(overrides=config)
+local = Context(config=invoke_config)
+
+
+def check_az_cli() -> None:
+    """Assert that the `az` CLI is installed and logged in."""
+    # E.g. on Ubuntu: `curl -sL https://aka.ms/InstallAzureCLIDeb | sudo bash`
+    assert local.run("az --version", warn=True), "Please install the `az` CLI!"
+    # TODO: Login with service principal (az login) and set
+    # default subscription (az account set -s) using secrets.
+    account: Result = local.run("az account show")
+    assert account.ok, "Please `az login`!"
+    sub = json.loads(account.stdout)
+    assert sub["isDefault"], "Please `az account set -s <subscription>`!"
+    logging.info(
+        f"Using account '{sub['user']['name']}' with subscription '{sub['name']}'"
+    )
+
+
+def create_boot_storage(location: str) -> str:
+    """Create a separate resource group and storage account for boot diagnostics."""
+    account = "pytestbootdiag"
+    # This command always exits with 0 but returns a string.
+    if local.run("az group exists -n pytest-lisa").stdout.strip() == "false":
+        local.run(f"az group create -n pytest-lisa --location {location}")
+    if not local.run(f"az storage account show -g pytest-lisa -n {account}", warn=True):
+        local.run(f"az storage account create -g pytest-lisa -n {account}")
```

Comment:
I assume that things like `pytestbootdiag` and `pytest-lisa` are hardcoded here just as a POC, right?

Reply:
Indeed! They should actually be generated based off a session ID, and torn down after the session ends. While it's very easy to create a VM with `az` where boot diagnostics are enabled, it _does not_ automatically create the necessary storage account, so this was some quick logic to do so.